compiler/rustc_hir/src/definitions.rs

```diff
@@ -436,7 +443,24 @@ impl DefPathData {
         }
     }
 
-    pub fn as_symbol(&self) -> Symbol {
+    pub fn name(&self) -> DefPathDataName {
+        use self::DefPathData::*;
+        match *self {
+            TypeNs(name) | ValueNs(name) | MacroNs(name) | LifetimeNs(name) => {
+                DefPathDataName::Named(name)
+            }
+            // Note that this does not show up in user print-outs.
+            CrateRoot => DefPathDataName::Anon { namespace: kw::Crate },
+            Impl => DefPathDataName::Anon { namespace: kw::Impl },
+            Misc => DefPathDataName::Anon { namespace: sym::misc },
+            ClosureExpr => DefPathDataName::Anon { namespace: sym::closure },
+            Ctor => DefPathDataName::Anon { namespace: sym::constructor },
+            AnonConst => DefPathDataName::Anon { namespace: sym::constant },
+            ImplTrait => DefPathDataName::Anon { namespace: sym::opaque },
+        }
+    }
+
+    pub fn as_legacy_symbol(&self) -> Symbol {
```

Comment:
Not super happy about keeping this around, it would be nice to not have `Symbol`s for the double-braced forms - they're not *really* needed for anything.

Reply:
What is better, calling `Symbol::intern` in the 3 or 4 places that needs a legacy symbol or keeping the old symbols around? I don't have a preference either way.

Reply:
It seems better to call `Symbol::intern` in a situation that shouldn't even be hit.

You could even e.g. write `if let DefPathDataName::Anon {..} = data.name() { unreachable!() }` and run tests, for each in turn.